### PR TITLE
List delete files in Iceberg's $files metadata table

### LINKED ISSUE #16233: Iceberg $files table not show delete files
For iceberg v2 format, if query `$files` metadata table, e.g. 
`select * from tmp."test_table$files" where content =1 limit 10;`
it will return empty result, the reason is that `FilesTable` only iterates `DataFile`, however I think it needs to scan `DeleteFile` as well. 

https://github.com/trinodb/trino/blob/64b0b7d9bf5b0314e0de64a8cee946bfeb283d99/plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java#L179

https://github.com/trinodb/trino/pull/16232 for the fix.

### PR DESCRIPTION
**Pull request by nineinchnick**
## Description

Based on #16232 with review comments applied - avoid reflection and use column position to read data. I preserved the commit author.

Fixes #16233

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Iceberg
* Fix `$files` table not showing delete files for Iceberg v2 format. ({issue}`[16233]`)
```

### CODE REVIEW

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 1]
```diff
@@ -57,17 +58,35 @@
...
 import static io.trino.spi.type.VarbinaryType.VARBINARY;
 import static io.trino.spi.type.VarcharType.VARCHAR;
 import static java.util.Objects.requireNonNull;
+import static org.apache.iceberg.MetadataTableType.FILES;
+import static org.apache.iceberg.MetadataTableUtils.createMetadataTableInstance;
 
 public class FilesTable
         implements SystemTable
 {
+    public static final String CONTENT_COLUMN_NAME = "content";
```
Comment (wendigo) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:76]: This could be extracted to a separate commit (unrelated to the commit title)

**Review by wendigo — APPROVED** [review]

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java [hunk 2]
```diff
@@ -57,17 +58,35 @@
...
+    public static final String FILE_SIZE_IN_BYTES_COLUMN_NAME = "file_size_in_bytes";
+    public static final String COLUMN_SIZES_COLUMN_NAME = "column_sizes";
+    public static final String VALUE_COUNTS_COLUMN_NAME = "value_counts";
+    public static final String NULL_VALUE_COUNTS_COLUMN_NAME = "null_value_counts";
+    public static final String NAN_VALUE_COUNTS_COLUMN_NAME = "nan_value_counts";
+    public static final String LOWER_BOUNDS_COLUMN_NAME = "lower_bounds";
+    public static final String UPPER_BOUNDS_COLUMN_NAME = "upper_bounds";
+    public static final String KEY_METADATA_COLUMN_NAME = "key_metadata";
+    public static final String SPLIT_OFFSETS_COLUMN_NAME = "split_offsets";
+    public static final String EQUALITY_IDS_COLUMN_NAME = "equality_ids";
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java:89]: Why are these constants public?

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java [hunk 3]
```diff
@@ -1584,7 +1584,7 @@ public void testUpdateWithSortOrder()
             assertQuery(
                     "SELECT custkey, name, address, nationkey, phone, acctbal, mktsegment, comment FROM " + table.getName(),
                     "SELECT custkey, name, address, nationkey, phone, acctbal, mktsegment, substring(comment, 2) FROM customer");
-            for (Object filePath : computeActual("SELECT file_path from \"" + table.getName() + "$files\"").getOnlyColumnAsSet()) {
+            for (Object filePath : computeActual("SELECT file_path from \"" + table.getName() + "$files\" where content != 1").getOnlyColumnAsSet()) {
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergConnectorTest.java:1587]: Uppercase `where`.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java [hunk 4]
```diff
@@ -96,6 +97,13 @@ public void setUp()
         assertUpdate("DELETE FROM test_schema.test_table_with_dml WHERE _date = DATE '2022-02-02' AND _varchar = 'b2'", 1);
         assertUpdate("INSERT INTO test_schema.test_table_with_dml VALUES ('c3', DATE '2022-03-03'), ('d1', DATE '2022-04-04')", 2);
         assertQuery("SELECT count(*) FROM test_schema.test_table_with_dml", "VALUES 7");
+
+        assertUpdate("CREATE TABLE test_schema.test_table_with_delete (_bigint BIGINT, _date DATE) WITH (partitioning = ARRAY['_date'])");
+        assertUpdate("INSERT INTO test_schema.test_table_with_delete VALUES (0, CAST('2019-09-08' AS DATE)), (1, CAST('2019-09-09' AS DATE)), (2, CAST('2019-09-09' AS DATE))", 3);
+        assertUpdate("INSERT INTO test_schema.test_table_with_delete VALUES (3, CAST('2019-09-09' AS DATE)), (4, CAST('2019-09-10' AS DATE)), (5, CAST('2019-09-10' AS DATE))", 3);
+        assertUpdate("DELETE FROM test_schema.test_table_with_delete WHERE _bigint = 5", 1);
+        assertUpdate("DELETE FROM test_schema.test_table_with_delete WHERE _bigint = 2", 1);
+        assertQuery("SELECT count(*) FROM test_schema.test_table_with_delete", "VALUES 4");
```
Comment (ebyhr) [inline plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java:106]: Could you move these statements to `testFilesTableWithDelete` method?

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java (hunk 1 above)
Comment (nineinchnick) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java]: Done!

File: plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java (hunk 2 above)
Comment (nineinchnick) [reply plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/FilesTable.java]: No reason, changed to private.

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java (hunk 4 above)
Comment (nineinchnick) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java]: Done!

**Review by wendigo — APPROVED** [review]

File: plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java [hunk 5]
```diff
@@ -363,6 +364,23 @@ public void testFilesTable()
...
+    {
+        assertUpdate("CREATE TABLE test_schema.test_table_with_delete (_bigint BIGINT, _date DATE) WITH (partitioning = ARRAY['_date'])");
+        assertUpdate("INSERT INTO test_schema.test_table_with_delete VALUES (0, CAST('2019-09-08' AS DATE)), (1, CAST('2019-09-09' AS DATE)), (2, CAST('2019-09-09' AS DATE))", 3);
+        assertUpdate("INSERT INTO test_schema.test_table_with_delete VALUES (3, CAST('2019-09-09' AS DATE)), (4, CAST('2019-09-10' AS DATE)), (5, CAST('2019-09-10' AS DATE))", 3);
+        assertUpdate("DELETE FROM test_schema.test_table_with_delete WHERE _bigint = 5", 1);
+        assertUpdate("DELETE FROM test_schema.test_table_with_delete WHERE _bigint = 2", 1);
+
+        assertQuery("SELECT count(*) FROM test_schema.test_table_with_delete", "VALUES 4");
+        assertQuery("SELECT count(*) FROM test_schema.\"test_table_with_delete$files\" WHERE content = " + FileContent.DATA.id(), "VALUES 4");
+        assertQuery("SELECT count(*) FROM test_schema.\"test_table_with_delete$files\" WHERE content = " + FileContent.POSITION_DELETES.id(), "VALUES 2");
```
[the remark this answers is not in the document]
Comment (nineinchnick) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java:378]: I could, but I thought we only want to have basic smoke tests as product tests. Is there any risk of this behaving differently in a full server?

[the remark this answers is not in the document]
Comment (wendigo) [reply plugin/trino-iceberg/src/test/java/io/trino/plugin/iceberg/BaseIcebergSystemTables.java:378]: I don't think we need a product test for that.
